lib/commands/touch.js

```diff
@@ -1,9 +1,99 @@
 commands.performActions = async function (actions) {
-  return await this.winAppDriver.sendCommand('/actions', 'POST', {actions});
+  //Workaround for https://github.com/appium/appium/issues/16268
+  //Once WAD offers full W3C support for keyboard and mouse type pointer actions, this can be proxied directly to the /actions endpoint
+  for (const command in actions) {
+    if (actions[command].type === 'key') {
+      this.log.info('Mapping keyboard actions');
+      return await performKeyboardCommands.bind(this)(actions[command].actions);
+    } else if (actions[command].type === 'pointer' && actions[command].parameters.pointerType === 'mouse') {
+      this.log.info('Mapping mouse actions');
+      return await performMouseCommands.bind(this)(actions[command].actions);
+    } else {
+      this.log.info('Going default actions route');
+      return await this.winAppDriver.sendCommand('/actions', 'POST', {actions});
+    }
+  }
 };
 
+/*
+  Supported actions:
+  -> sendKeys(CharSequence... keys)
+  -> sendKeysInTicks(CharSequence... keys)
+*/
+// TODO: remove once WAD offers W3C for keyboard input actions
+async function performKeyboardCommands(actions) {
+  const keysFormatted = formatKeyboardInput(actions);
+  return await this.winAppDriver.sendCommand('/keys', 'POST', {'value': keysFormatted });
+}
+
+/*
+  Supported actions:
+  -> click()
+  -> click(WebElement target)
+  -> contextClick()
+  -> contextClick(WebElement target)
+  -> release()
+  -> release(WebElement target)
+  -> clickAndHold()
+  -> clickAndHold(WebElement target)
+  -> doubleClick()
+  -> doubleClick(WebElement target)
+  -> moveToElement(WebElement target)
+  -> moveToElement(WebElement target, int xOffset, int yOffset)
+  -> dragAndDrop(WebElement source, WebElement target)
+  -> pause()
+
+*/
+// TODO: remove once WAD offers W3C support for pointer actions of type mouse
+async function performMouseCommands(actions) {
+  let result;
+  for (const action in actions) {
+    switch (actions[action].type) {
+      case 'pointerMove':
+        this.log.info(JSON.stringify(actions[action]));
+        this.log.info('Done');
+        result = await this.winAppDriver.sendCommand('/moveto', 'POST', getExpectedFormatForMouseMovement(actions[action]));
+        break;
+      case 'pointerDown':
+        result = await this.winAppDriver.sendCommand('/buttondown', 'POST', {'button': actions[action].button });
+        break;
+      case 'pointerUp':
+        result = await this.winAppDriver.sendCommand('/buttonup', 'POST', {'button': actions[action].button });
+        break;
+      case 'pause':
+        await sleep(actions[action].duration);
+        break;
+      default:
+        this.log.info('No mouse action mapped for type: ${JSON.stringify(actions[action].type)}, moving to next action');
+    }
+  };
+  return result;
+}
+
+function formatKeyboardInput(actions) {
+  const keys = new Array();
+  actions.forEach((action) => {
```

Comment:
I would rather look for keyDown/keyUp pairs

Reply:
the actions.sendKeys() command create an array of actions that basically does a keyDown and a keyUp for each character
so if we send ("ab") chars the actions array will contains 4 actions
(keyDown a, keyDown b, keyUp a, keyUp b)

I basically need just the chars value themselves, that's the body the /keys endpoint from WAD accepts because he takes care of the keyUp and KeyDown actions itself. So I need "ab"
I've been adding all the values under the keyDown actions. They always come as pair true, so I didn't see a point of checking keyUp too. I could do some extra checks that number of actions for each match or that they contain the same value but it seems a bit redundant.
Making sure each char gets it's keyUp/keyDown value should be appium's responsability

Also, we don't support just a key press either up or down...there's no endpoint from WAD that does it :(

Reply:
I completely understand your point. Although I would rather prefer to be strict on what we can transform and what we cannot. For example, if an action simply presses some key and leaves it be in that state then our transformation won't be correct. same situation if some key is depressed first and then pressed again. We must throw an explicit error that we are unable to parse such action chains rather than (erroneously) trying to apply them